public/app/features/dashboard/state/DashboardModel.ts

```diff
@@ -893,6 +902,14 @@ export class DashboardModel {
           // update insert post and y max
           insertPos += 1;
           yMax = Math.max(yMax, panel.gridPos.y + panel.gridPos.h);
+
+          const sharedDashboardPanels = this.panels.filter(
+            (p) => p.datasource === SHARED_DASHBOARD_QUERY && p.targets.some((t: any) => t.panelId === panel.id)
+          );
+
+          for (const sharedPanel of sharedDashboardPanels) {
+            sharedPanel.refresh();
+          }
```

Comment:
Can you describe this bit? panels will automatically refresh rows are expanded (on mount)

Reply:
The other logic introduced in this PR makes it so if getPanelById(id) is called, but the panel with that ID is only found in a collapsed row, a new panel model is made, which in the case of the bug that this PR is solving will be used for the dashboard datasource panel.
If the code above wasn't included, when the collapsed row is expanded another panel model would be created, and if any changes were made to it they wouldn't be reflected in the dashboard datasource panel. So the code is there to refresh any dashboard datasource panels pointing to a panel included in the row being expanded, so that the source and target refer to the same panel model.